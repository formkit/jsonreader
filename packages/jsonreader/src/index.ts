import JSON5 from "json5";
import { eq } from "@formkit/utils";

/**
 * Options for controlling how the JSON reader yields progress
 */
export interface JsonReaderOptions {
  /**
   * Required properties that must be present before yielding a result
   */
  required?: string[];
  
  /**
   * Properties to exclude from partial results
   */
  silent?: string[];
  
  /**
   * Additional values to assign to each yielded result
   */
  assign?: Record<string, any>;
}

/**
 * An async generator function that reads the stream as it arrives and yields a progressively updated input until it
 * is complete.
 * @param reader - ReadableStreamDefaultReader<Uint8Array>
 * @param options - Options for how to yield the progress.
 */
export async function* jsonReader(
  reader: ReadableStreamDefaultReader<Uint8Array>,
  options?: JsonReaderOptions
) {
  const decoder = new TextDecoder();
  const buffer: string[] = [];
  const propertyPath: string[] = [];
  const depthMap: Array<"{" | "["> = [];
  let readingProperty = false;
  let willBeReadingProperty = false;
  // let readingValue = false;
  let currentProperty = "";
  let currentValue = "";
  let idx = 0;
  let lastChar = "";
  let isQuoted = false;
  let requiredPropertiesCompleted: boolean | undefined =
    !options?.required?.length;
  let completedProperties: string[] = [];
  let propertyComplete = false;
  let lastYield = {};
  let didYield = false;
  let isEscaped = false;

  const parentIsObject = () => depthMap[depthMap.length - 1] === "{";
  const parentIsArray = () => depthMap[depthMap.length - 1] === "[";

  for await (const chunk of read(reader)) {
    buffer.push(...[...decoder.decode(chunk)]);
    while (idx < buffer.length) {
      const char = buffer[idx++];
      isEscaped = lastChar === "\\";
      lastChar = char;

      if (willBeReadingProperty && char.trim() !== "") {
        readingProperty = true;
        willBeReadingProperty = false;
      }

      if (propertyComplete && char.trim() !== "") {
        readingProperty = false;
        propertyComplete = false;
      }
      // Skip escaped characters
      if (!isEscaped && (!isQuoted || char === '"')) {
        switch (char) {
          case '"':
            isQuoted = !isQuoted;
            break;
          case "{":
            currentProperty = "";
            willBeReadingProperty = true;
            depthMap.push(char);
            continue;
          case "[":
            readingProperty = false;
            depthMap.push(char);
            propertyPath.push("0");
            continue;
          case ":":
            propertyComplete = true;
            currentValue = "";
            propertyPath.push(trim(currentProperty, /["' \n]/));
            continue;
          case "}":
          case "]":
            depthMap.pop();
            currentProperty = "";
            currentValue = "";
            completedProperties.push(propertyPath.join("."));
            propertyPath.pop();
            continue;
          case ",":
            if (parentIsObject() && !readingProperty) {
              willBeReadingProperty = true;
              currentProperty = "";
              completedProperties.push(propertyPath.join("."));
              currentValue = "";
              propertyPath.pop();
            } else if (parentIsArray()) {
              const newIndex =
                parseInt(propertyPath[propertyPath.length - 1]) + 1;
              propertyPath.pop();
              propertyPath.push(newIndex.toString());
              completedProperties.push(propertyPath.join("."));
              currentValue = "";
            }
            continue;
        }
      }
      if (readingProperty) {
        currentProperty += char;
      } else if (isQuoted || !/\s/.test(char) || isBool(currentValue, char)) {
        currentValue += char;
      }
    }
    if (isEscaped) {
      // We dont track the "level" of the escape, so we skip it.
      continue;
    }
    if (isBool(currentValue, "", false)) {
      // If it is an incomplete boolean, like "tru" or "fal" we skip it.
      continue;
    }
    // Determine whether or not to yield a chunk. We should yield if
    // 1. All the required properties have been completed
    // 2. Any new properties were completed
    // 3. We are not processing the property of an object
    // 4. The current property being read is not "silent"
    if (!requiredPropertiesCompleted) {
      requiredPropertiesCompleted = options?.required?.every((prop) =>
        completedProperties.includes(prop)
      );
    }

    if (propertyComplete && currentValue.trim() === "") {
      // We are at a position like: `{ type: `, so we cannot parse the value yet.
      continue;
    }

    let currentSilentPropertyPath: string[] | false = false;
    if (options?.silent) {
      for (const silentPath of options.silent) {
        if (isSilentPath(propertyPath, silentPath)) {
          currentSilentPropertyPath = propertyPath.slice(
            0,
            silentPath.split(".").length
          );
        }
      }
    }

    if (requiredPropertiesCompleted && !readingProperty) {
      let incompleteJSON = buffer.join("");
      if (incompleteJSON.trim() === "") {
        continue;
      }
      if (isQuoted) {
        if (lastChar === "\\") {
          incompleteJSON = incompleteJSON.slice(0, -1);
        }
        incompleteJSON += '"';
      }
      for (let i = depthMap.length - 1; i >= 0; i--) {
        incompleteJSON += depthMap[i] === "{" ? "}" : "]";
      }

      try {
        let parsedValue = Object.assign(
          {},
          options?.assign ?? {},
          JSON5.parse(incompleteJSON)
        );

        // Delete the silent properties that are not yet completed.
        if (currentSilentPropertyPath) {
          let current = parsedValue;
          for (const [index, segment] of currentSilentPropertyPath.entries()) {
            if (index < currentSilentPropertyPath.length - 1) {
              current = current[segment];
            } else {
              delete current[segment];
            }
          }
        }

        parsedValue = removeEmptyValues(parsedValue);
        if (!eq(parsedValue, lastYield)) {
          lastYield = parsedValue;
          didYield = true;
          yield parsedValue;
        }
      } catch (err) {
        console.warn(
          `Could not parse JSON: ${incompleteJSON}, buffer: ${buffer.join()}`
        );
      }
    }
  }
  if (!didYield && buffer.join("").trim() !== "") {
    yield Object.assign(
      {},
      options?.assign ?? {},
      JSON5.parse(buffer.join(""))
    );
  }
}

/**
 * Checks if a property path should be silenced based on a silent path pattern
 * 
 * @param propertyPath - Current property path as an array of segments
 * @param silentPath - Path pattern to check against (supports * as wildcard)
 * @returns True if the property should be silenced
 */
function isSilentPath(propertyPath: string[], silentPath: string) {
  const silentSegments = silentPath.split(".");
  const currentSegments = propertyPath;
  return (
    silentSegments.length <= currentSegments.length &&
    silentSegments.every((segment, index) => {
      if (segment === "*") return true;
      return segment === currentSegments[index];
    })
  );
}

/**
 * Removes empty values from an object or array recursively.
 * @param obj - The object or array to clean.
 * @returns A new object or array with empty values removed.
 */
export function removeEmptyValues<
  T extends Record<string, unknown> | unknown[] | unknown
>(obj: T): T {
  if (typeof obj !== "object" || obj === null) {
    return obj;
  }

  if (Array.isArray(obj)) {
    return obj
      .map(removeEmptyValues)
      .filter((v) => v !== null && v !== undefined) as T;
  }

  const result: Record<string, any> = {};
  for (const key in obj) {
    const value = removeEmptyValues(obj[key]);
    if (value !== null && value !== undefined) {
      result[key] = value;
    }
  }
  return result as T;
}

/**
 * Read the stream as JSON, but only yields values at the given paths when they are complete.
 * @param reader - The reader to read from
 * @param paths - An array of dot-notation paths (where * can be used as a wildcard)
 */
export async function* jsonPathReader<Paths extends string[]>(
  reader: ReadableStreamDefaultReader<Uint8Array>,
  paths: Paths
): AsyncGenerator<[any, Paths[number]]> {
  if (!paths?.length) {
    return;
  }
  const ignoredProperties = new Set<string>();

  // Helper to check if a path is a subpath of any target path
  const isSubpathOfTarget = (path: string) => {
    return paths.some((targetPath) => {
      const targetSegments = targetPath.split(".");
      const pathSegments = path.split(".");
      return (
        pathSegments.length < targetSegments.length &&
        targetSegments.slice(0, pathSegments.length).join(".") === path
      );
    });
  };
  
  // Process each chunk of JSON data as it arrives
  for await (const chunk of jsonReader(reader, { silent: paths })) {
    // Stack for iterative traversal of nested objects/arrays
    const stack: Array<{ obj: any; path: string }> = [{ obj: chunk, path: "" }];

    while (stack.length > 0) {
      const { obj, path } = stack.pop()!;

      if (typeof obj !== "object" || obj === null) continue;

      for (const key in obj) {
        const fullPath = path ? `${path}.${key}` : key;

        if (ignoredProperties.has(fullPath)) {
          delete obj[key];
          continue;
        }

        const value = obj[key];

        for (const p of paths) {
          if (isSilentPath(fullPath.split("."), p)) {
            yield [value, p] as const;
            ignoredProperties.add(fullPath);
            delete obj[key];
            break;
          }
        }

        if (!ignoredProperties.has(fullPath)) {
          if (!isSubpathOfTarget(fullPath)) {
            ignoredProperties.add(fullPath);
            delete obj[key];
          } else if (typeof value === "object" && value !== null) {
            stack.push({ obj: value, path: fullPath });
          }
        }
      }
    }
  }
}

/**
 * Check if the current value is adding up to a boolean.
 * @param currentValue - The current value being read.
 * @param char - The current character being read.
 * @param allowComplete - Whether to allow complete boolean values
 * @returns True if the current value is a boolean, false otherwise.
 */
function isBool(currentValue: string, char: string, allowComplete = true) {
  const partial = `${currentValue.trim()}${char}`;
  if (!partial) return false;
  if (!allowComplete && (partial === "true" || partial === "false")) {
    return false;
  }
  return "true".startsWith(partial) || "false".startsWith(partial);
}

/**
 * An async generator to read the stream one chunk at a time.
 * @param reader - ReadableStreamDefaultReader<Uint8Array>
 */
export async function* read<T>(
  reader: ReadableStreamDefaultReader<T>
): AsyncGenerator<T> {
  let value: T | undefined = undefined;
  let done = false;
  do {
    ({ value, done } = await reader.read());
    if (value !== undefined) yield value;
  } while (!done);
}

/**
 * Trim a string by removing specified characters
 * @param s - String to trim
 * @param chars - Regular expression of characters to trim
 * @returns Trimmed string
 */
export function trim(s: string, chars: RegExp): string {
  // Trim from the start
  let start = 0;
  while (start < s.length && chars.test(s[start])) {
    start++;
  }

  // Trim from the end
  let end = s.length - 1;
  while (end > start && chars.test(s[end])) {
    end--;
  }

  return s.slice(start, end + 1);
}

export default {
  jsonReader,
  jsonPathReader,
  removeEmptyValues,
  read,
  trim
}; 