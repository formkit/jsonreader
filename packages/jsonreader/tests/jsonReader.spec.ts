import { describe, it, vi } from "vitest";
import { createStream } from "./helpers/streamHelpers";
import { jsonReader, jsonPathReader } from "../src";

describe("streaming json", () => {
  it("doesnt warn when empty", async ({ expect }) => {
    const consoleSpy = vi.spyOn(console, "warn").mockImplementation(() => {});
    const stream = createStream(" ");
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      required: [],
      silent: undefined,
      assign: { foo: "bar" },
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`[]`);
    expect(consoleSpy).not.toHaveBeenCalled();
    consoleSpy.mockRestore();
  });
  it("can yield an empty object", async ({ expect }) => {
    const stream = createStream("{}");
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      required: [],
      silent: undefined,
      assign: { foo: "bar" },
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": "bar",
        },
      ]
    `);
  });
  it("can only yields when values are being read", async ({ expect }) => {
    const stream = createStream('{ "foo": "bar" }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader())) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": "",
        },
        {
          "foo": "b",
        },
        {
          "foo": "ba",
        },
        {
          "foo": "bar",
        },
      ]
    `);
  });

  it("only yields when values are read with multiple properties", async ({
    expect,
  }) => {
    const stream = createStream('{ "foo": "b", "cat": 123 }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader())) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": "",
        },
        {
          "foo": "b",
        },
        {
          "cat": 1,
          "foo": "b",
        },
        {
          "cat": 12,
          "foo": "b",
        },
        {
          "cat": 123,
          "foo": "b",
        },
      ]
    `);
  });

  it("only yields when required properties are complete", async ({
    expect,
  }) => {
    const stream = createStream('{ "foo": "bar", "baz": 123 }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      required: ["foo"],
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": "bar",
        },
        {
          "baz": 1,
          "foo": "bar",
        },
        {
          "baz": 12,
          "foo": "bar",
        },
        {
          "baz": 123,
          "foo": "bar",
        },
      ]
    `);
  });

  it("can yield array property values", async ({ expect }) => {
    const stream = createStream('{ "foo": ["one", "two", 123], "baz": [123] }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader())) {
      yields.push(obj);
    }
    expect(yields).toMatchSnapshot();
  });

  it("is fault tolerant to missing quote marks", async ({ expect }) => {
    const stream = createStream('{ foo: "bar" }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {})) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": "",
        },
        {
          "foo": "b",
        },
        {
          "foo": "ba",
        },
        {
          "foo": "bar",
        },
      ]
    `);
  });

  it("can have required properties that are arrays", async ({ expect }) => {
    const stream = createStream('{ foo: ["first", "last"], bar: 10 }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      required: ["foo"],
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": [
            "first",
            "last",
          ],
        },
        {
          "bar": 1,
          "foo": [
            "first",
            "last",
          ],
        },
        {
          "bar": 10,
          "foo": [
            "first",
            "last",
          ],
        },
      ]
    `);
  });
  it("can have required properties that are nested", async ({ expect }) => {
    const stream = createStream(
      '{ foo: { bim: "here", baz: 123 }, "hello": "world" }'
    );
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      required: ["foo.bim"],
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchSnapshot();
  });

  it("wont emit when reading a silent property value", async ({ expect }) => {
    const stream = createStream(
      '{ type: "str", "format": ["123", 465], "value": "1" }'
    );
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      required: ["type"],
      silent: ["format"],
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "type": "str",
        },
        {
          "format": [
            "123",
            465,
          ],
          "type": "str",
        },
        {
          "format": [
            "123",
            465,
          ],
          "type": "str",
          "value": "",
        },
        {
          "format": [
            "123",
            465,
          ],
          "type": "str",
          "value": "1",
        },
      ]
    `);
  });
  it("can handle escaped double quotes", async ({ expect }) => {
    const stream = createStream('{ "foo": "bar\\"baz\\"" }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader())) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": "",
        },
        {
          "foo": "b",
        },
        {
          "foo": "ba",
        },
        {
          "foo": "bar",
        },
        {
          "foo": "bar\"b",
        },
        {
          "foo": "bar\"ba",
        },
        {
          "foo": "bar\"baz",
        },
        {
          "foo": "bar\"baz\"",
        },
      ]
    `);
  });

  it("can be silent for all items in an array", async ({ expect }) => {
    const stream = createStream('{ "foo": [{f: 1}, {b: 2, c: 3}, "345"] }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      silent: ["foo.*"],
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "foo": [],
        },
        {
          "foo": [
            {
              "f": 1,
            },
          ],
        },
        {
          "foo": [
            {
              "f": 1,
            },
            {
              "b": 2,
              "c": 3,
            },
          ],
        },
        {
          "foo": [
            {
              "f": 1,
            },
            {
              "b": 2,
              "c": 3,
            },
            "345",
          ],
        },
      ]
    `);
  });

  it("does not return early when dealing with nested strings of json", async ({
    expect,
  }) => {
    const stream = createStream(
      `{"response":{"routes":[{"prompt_name":"routeEditInput","instructions_for_prompt":"Translate this input to German","arguments_for_prompt":[{"key":"rewritten_prompt","value":"Translate this input to German"},{"key":"rewritten_input","value":"{\\"type\\":\\"page_break\\",\\"idx\\":\\"1\\",\\"key\\":\\"fvumampbuem\\",\\"content\\":\\"$: \\\\\\"User Login\\\\\\"\\"}"}]}]}}`
    );
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader(), {
      silent: ["response.routes.*"],
    })) {
      yields.push(obj);
    }
    expect(yields).toMatchSnapshot();
  });

  it("does not yield incomplete boolean values", async ({ expect }) => {
    const stream = createStream('{ "isComplete": true }');
    const yields: any[] = [];
    for await (const obj of jsonReader(stream.getReader())) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        {
          "isComplete": true,
        },
      ]
    `);
  });
});

describe("jsonPathReader", () => {
  it("yields values when array items match wildcard paths", async ({
    expect,
  }) => {
    const stream = createStream(`
      {
        "items": [
          {
            "id": 1,
            "data": "first"
          },
          {
            "id": 2,
            "data": "second"
          },
          {
            "id": 3,
            "data": "third"
          }
        ]
      }
    `);

    const yields: any[] = [];
    for await (const obj of jsonPathReader(stream.getReader(), ["items.*"])) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        [
          {
            "data": "first",
            "id": 1,
          },
          "items.*",
        ],
        [
          {
            "data": "second",
            "id": 2,
          },
          "items.*",
        ],
        [
          {
            "data": "third",
            "id": 3,
          },
          "items.*",
        ],
      ]
    `);
  });

  it("yields values for deep properties and root properties", async ({
    expect,
  }) => {
    const stream = createStream(`{
        "rootProp": "root value",
        "nested": {
          "deep": {
            "property": "deep value",
            "other": "ignored"
          },
          "ignored": true
        },
        "alsoIgnored": false
      }`);

    const yields: any[] = [];
    for await (const obj of jsonPathReader(stream.getReader(), [
      "rootProp",
      "nested.deep.property",
    ])) {
      yields.push(obj);
    }
    expect(yields).toMatchInlineSnapshot(`
      [
        [
          "root value",
          "rootProp",
        ],
        [
          "deep value",
          "nested.deep.property",
        ],
      ]
    `);
  });
});
