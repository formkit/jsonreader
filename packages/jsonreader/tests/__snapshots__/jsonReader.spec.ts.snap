// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`streaming json > can have required properties that are nested 1`] = `
[
  {
    "foo": {
      "bim": "here",
    },
  },
  {
    "foo": {
      "baz": 1,
      "bim": "here",
    },
  },
  {
    "foo": {
      "baz": 12,
      "bim": "here",
    },
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
    "hello": "",
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
    "hello": "w",
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
    "hello": "wo",
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
    "hello": "wor",
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
    "hello": "worl",
  },
  {
    "foo": {
      "baz": 123,
      "bim": "here",
    },
    "hello": "world",
  },
]
`;

exports[`streaming json > can yield array property values 1`] = `
[
  {
    "foo": [],
  },
  {
    "foo": [
      "",
    ],
  },
  {
    "foo": [
      "o",
    ],
  },
  {
    "foo": [
      "on",
    ],
  },
  {
    "foo": [
      "one",
    ],
  },
  {
    "foo": [
      "one",
      "",
    ],
  },
  {
    "foo": [
      "one",
      "t",
    ],
  },
  {
    "foo": [
      "one",
      "tw",
    ],
  },
  {
    "foo": [
      "one",
      "two",
    ],
  },
  {
    "foo": [
      "one",
      "two",
      1,
    ],
  },
  {
    "foo": [
      "one",
      "two",
      12,
    ],
  },
  {
    "foo": [
      "one",
      "two",
      123,
    ],
  },
  {
    "baz": [],
    "foo": [
      "one",
      "two",
      123,
    ],
  },
  {
    "baz": [
      1,
    ],
    "foo": [
      "one",
      "two",
      123,
    ],
  },
  {
    "baz": [
      12,
    ],
    "foo": [
      "one",
      "two",
      123,
    ],
  },
  {
    "baz": [
      123,
    ],
    "foo": [
      "one",
      "two",
      123,
    ],
  },
]
`;

exports[`streaming json > does not return early when dealing with nested strings of json 1`] = `
[
  {
    "response": {},
  },
  {
    "response": {
      "routes": [],
    },
  },
  {
    "response": {
      "routes": [
        {
          "arguments_for_prompt": [
            {
              "key": "rewritten_prompt",
              "value": "Translate this input to German",
            },
            {
              "key": "rewritten_input",
              "value": "{"type":"page_break","idx":"1","key":"fvumampbuem","content":"$: \\"User Login\\""}",
            },
          ],
          "instructions_for_prompt": "Translate this input to German",
          "prompt_name": "routeEditInput",
        },
      ],
    },
  },
]
`;
